File: scripts/generate-icons.py
#!/usr/bin/env python3
"""Generate WorldCapIQ brand icons (PNG + ICO) from spec."""

import os
import sys

try:
    from PIL import Image, ImageDraw, ImageFont
except ImportError:
    print("Installing Pillow...")
    os.system(f"{sys.executable} -m pip install Pillow")
    from PIL import Image, ImageDraw, ImageFont

# Paths
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.dirname(SCRIPT_DIR)
FONT_PATH = os.path.join(PROJECT_ROOT, "public", "fonts", "bebas-neue-400.ttf")
ICONS_DIR = os.path.join(PROJECT_ROOT, "public", "icons")
PUBLIC_DIR = os.path.join(PROJECT_ROOT, "public")

# Brand colors
BG_GREEN = (160, 212, 148)       # #a0d494
TEXT_DARK = (13, 15, 13)          # #0d0f0d
ACCENT_LINE = (13, 15, 13, 102)  # #0d0f0d at 40% opacity

os.makedirs(ICONS_DIR, exist_ok=True)


def draw_squircle(draw, size, radius, fill):
    """Draw a rounded rectangle (squircle approximation)."""
    draw.rounded_rectangle([0, 0, size - 1, size - 1], radius=radius, fill=fill)


def draw_accent_line(draw, size):
    """Draw diagonal accent line in bottom-right."""
    x_start = int(size * 0.625)
    y_start = size
    x_end = size
    y_end = int(size * 0.625)
    line_width = max(1, int(size * 0.006))
    draw.line([(x_start, y_start), (x_end, y_end)], fill=ACCENT_LINE, width=line_width)


def create_standard_icon(size, with_accent=True):
    """Create a standard 'any' purpose icon at given size."""
    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    # Squircle background (22% radius)
    radius = int(size * 0.22)
    draw_squircle(draw, size, radius, BG_GREEN)

    # Accent line
    if with_accent and size >= 64:
        draw_accent_line(draw, size)

    # IQ text
    font_size = int(size * 0.62)
    try:
        font = ImageFont.truetype(FONT_PATH, font_size)
    except (OSError, IOError):
        print(f"  Warning: Could not load Bebas Neue, using default font")
        font = ImageFont.load_default()

    text = "IQ"
    bbox = draw.textbbox((0, 0), text, font=font)
    text_w = bbox[2] - bbox[0]
    text_h = bbox[3] - bbox[1]

    # Center the text (adjust y slightly up for visual centering)
    x = (size - text_w) // 2 - bbox[0]
    y = (size - text_h) // 2 - bbox[1] - int(size * 0.02)

    draw.text((x, y), text, fill=TEXT_DARK, font=font)
    return img


def create_maskable_icon(size):
    """Create a maskable icon (full-bleed, content in safe zone)."""
    img = Image.new("RGBA", (size, size), BG_GREEN + (255,))
    draw = ImageDraw.Draw(img)

    # Content in safe zone (inner 80%)
    safe_zone = int(size * 0.8)
    offset = int(size * 0.1)

    # IQ text sized to safe zone
    font_size = int(safe_zone * 0.62)
    try:
        font = ImageFont.truetype(FONT_PATH, font_size)
    except (OSError, IOError):
        font = ImageFont.load_default()

    text = "IQ"
    bbox = draw.textbbox((0, 0), text, font=font)
    text_w = bbox[2] - bbox[0]
    text_h = bbox[3] - bbox[1]

    x = (size - text_w) // 2 - bbox[0]
    y = (size - text_h) // 2 - bbox[1] - int(size * 0.02)

    draw.text((x, y), text, fill=TEXT_DARK, font=font)
    return img


def create_favicon_ico(icon_img):
    """Create a multi-resolution favicon.ico."""
    sizes = [(16, 16), (32, 32), (48, 48)]
    ico_images = []
    for s in sizes:
        resized = icon_img.resize(s, Image.LANCZOS)
        ico_images.append(resized)
    return ico_images


def main():
    print("Generating WorldCapIQ brand icons...")
    print(f"  Font: {FONT_PATH}")
    print(f"  Output: {ICONS_DIR}")
    print()

    # Standard icons (purpose: "any")
    standard_sizes = {
        "icon-192.png": 192,
        "icon-512.png": 512,
    }

    for filename, size in standard_sizes.items():
        path = os.path.join(ICONS_DIR, filename)
        img = create_standard_icon(size)
        img.save(path, "PNG", optimize=True)
        file_size = os.path.getsize(path)
        print(f"  Created {filename} ({size}x{size}, {file_size:,} bytes)")

    # Maskable icons (purpose: "maskable")
    maskable_sizes = {
        "icon-maskable-192.png": 192,
        "icon-maskable-512.png": 512,
    }

    for filename, size in maskable_sizes.items():
        path = os.path.join(ICONS_DIR, filename)
        img = create_maskable_icon(size)
        img.save(path, "PNG", optimize=True)
        file_size = os.path.getsize(path)
        print(f"  Created {filename} ({size}x{size}, {file_size:,} bytes)")

    # Apple touch icon (180x180, standard style)
    apple_icon = create_standard_icon(180)
    apple_path = os.path.join(PUBLIC_DIR, "apple-touch-icon.png")
    apple_icon.save(apple_path, "PNG", optimize=True)
    file_size = os.path.getsize(apple_path)
    print(f"  Created apple-touch-icon.png (180x180, {file_size:,} bytes)")

    # Favicon.ico (multi-res: 16, 32, 48)
    base_icon = create_standard_icon(512)
    ico_path = os.path.join(PUBLIC_DIR, "favicon.ico")
    ico_16 = base_icon.resize((16, 16), Image.LANCZOS)
    ico_32 = base_icon.resize((32, 32), Image.LANCZOS)
    ico_48 = base_icon.resize((48, 48), Image.LANCZOS)
    ico_16.save(ico_path, format="ICO", sizes=[(16, 16), (32, 32), (48, 48)],
                append_images=[ico_32, ico_48])
    file_size = os.path.getsize(ico_path)
    print(f"  Created favicon.ico (16+32+48, {file_size:,} bytes)")

    print()
    print("Done. All icons generated.")


if __name__ == "__main__":
    main()

File: scripts/generate-icons.mjs
#!/usr/bin/env node
/**
 * Generate WorldCapIQ brand icons (PNG + ICO) from SVG.
 * Uses sharp (already a project dependency).
 */

import sharp from 'sharp';
import { writeFileSync, mkdirSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = join(__dirname, '..');
const ICONS_DIR = join(PROJECT_ROOT, 'public', 'icons');
const PUBLIC_DIR = join(PROJECT_ROOT, 'public');

// Brand colors
const BG_GREEN = '#a0d494';
const TEXT_DARK = '#0d0f0d';

mkdirSync(ICONS_DIR, { recursive: true });

/**
 * Create an SVG string for the IQ icon mark at a given size.
 * Standard variant: squircle with rounded corners.
 */
function createStandardIconSvg(size) {
  const radius = Math.round(size * 0.22);
  const fontSize = Math.round(size * 0.58);
  const lineWidth = Math.max(1, Math.round(size * 0.006));
  const lineX1 = Math.round(size * 0.625);
  const lineY2 = Math.round(size * 0.625);
  // Shift text up slightly for visual centering (Bebas Neue has high cap height)
  const textY = Math.round(size * 0.66);

  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${size} ${size}" width="${size}" height="${size}">
  <defs>
    <clipPath id="sq">
      <rect width="${size}" height="${size}" rx="${radius}" ry="${radius}"/>
    </clipPath>
  </defs>
  <rect width="${size}" height="${size}" rx="${radius}" ry="${radius}" fill="${BG_GREEN}"/>
  <line x1="${lineX1}" y1="${size}" x2="${size}" y2="${lineY2}" stroke="${TEXT_DARK}" stroke-width="${lineWidth}" opacity="0.4" clip-path="url(#sq)"/>
  <text x="${size / 2}" y="${textY}" text-anchor="middle" font-family="Impact,'Arial Narrow',Helvetica,sans-serif" font-size="${fontSize}" font-weight="400" fill="${TEXT_DARK}" letter-spacing="-${Math.round(size * 0.015)}">IQ</text>
</svg>`;
}

/**
 * Maskable variant: full-bleed green, no rounded corners, content in safe zone.
 */
function createMaskableIconSvg(size) {
  const fontSize = Math.round(size * 0.8 * 0.58);
  const textY = Math.round(size * 0.64);

  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${size} ${size}" width="${size}" height="${size}">
  <rect width="${size}" height="${size}" fill="${BG_GREEN}"/>
  <text x="${size / 2}" y="${textY}" text-anchor="middle" font-family="Impact,'Arial Narrow',Helvetica,sans-serif" font-size="${fontSize}" font-weight="400" fill="${TEXT_DARK}" letter-spacing="-${Math.round(size * 0.012)}">IQ</text>
</svg>`;
}

async function generateIcon(name, size, svgFn, outputDir) {
  const svg = svgFn(size);
  const buffer = await sharp(Buffer.from(svg))
    .resize(size, size)
    .png({ compressionLevel: 9 })
    .toBuffer();

  const path = join(outputDir, name);
  writeFileSync(path, buffer);
  const fileSize = statSync(path).size;
  console.log(`  ${name} (${size}x${size}, ${fileSize.toLocaleString()} bytes)`);
  return buffer;
}

async function generateFaviconIco(svg512Buffer) {
  // Create 16x16, 32x32, and 48x48 PNGs
  const sizes = [16, 32, 48];
  const pngBuffers = await Promise.all(
    sizes.map(s =>
      sharp(svg512Buffer).resize(s, s).png({ compressionLevel: 9 }).toBuffer()
    )
  );

  // Build a minimal ICO file
  // ICO format: header (6 bytes) + directory entries (16 bytes each) + image data
  const header = Buffer.alloc(6);
  header.writeUInt16LE(0, 0);      // Reserved
  header.writeUInt16LE(1, 2);      // ICO type
  header.writeUInt16LE(sizes.length, 4); // Number of images

  let dataOffset = 6 + (16 * sizes.length);
  const dirEntries = [];
  for (let i = 0; i < sizes.length; i++) {
    const entry = Buffer.alloc(16);
    entry.writeUInt8(sizes[i] === 256 ? 0 : sizes[i], 0);  // Width
    entry.writeUInt8(sizes[i] === 256 ? 0 : sizes[i], 1);  // Height
    entry.writeUInt8(0, 2);           // Color palette
    entry.writeUInt8(0, 3);           // Reserved
    entry.writeUInt16LE(1, 4);        // Color planes
    entry.writeUInt16LE(32, 6);       // Bits per pixel
    entry.writeUInt32LE(pngBuffers[i].length, 8);  // Image size
    entry.writeUInt32LE(dataOffset, 12);            // Image offset
    dirEntries.push(entry);
    dataOffset += pngBuffers[i].length;
  }

  return Buffer.concat([header, ...dirEntries, ...pngBuffers]);
}

async function main() {
  console.log('Generating WorldCapIQ brand icons...\n');

  // Standard icons
  console.log('Standard icons (purpose: "any"):');
  await generateIcon('icon-192.png', 192, createStandardIconSvg, ICONS_DIR);
  const icon512Svg = Buffer.from(createStandardIconSvg(512));
  await generateIcon('icon-512.png', 512, createStandardIconSvg, ICONS_DIR);

  // Maskable icons
  console.log('\nMaskable icons (purpose: "maskable"):');
  await generateIcon('icon-maskable-192.png', 192, createMaskableIconSvg, ICONS_DIR);
  await generateIcon('icon-maskable-512.png', 512, createMaskableIconSvg, ICONS_DIR);

  // Apple touch icon
  console.log('\nApple touch icon:');
  await generateIcon('apple-touch-icon.png', 180, createStandardIconSvg, PUBLIC_DIR);

  // Favicon.ico
  console.log('\nFavicon:');
  const icoBuffer = await generateFaviconIco(icon512Svg);
  const icoPath = join(PUBLIC_DIR, 'favicon.ico');
  writeFileSync(icoPath, icoBuffer);
  const icoSize = statSync(icoPath).size;
  console.log(`  favicon.ico (16+32+48, ${icoSize.toLocaleString()} bytes)`);

  console.log('\nDone. All icons generated.');
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});
